openedx_learning/apps/core/itemstore/models.py

```diff
@@ -0,0 +1,133 @@
+    # we don't currently have the data integrity guarantees to make sure that 
+    # multiple versions aren't active at the same time. Maybe add a ref to the
+    # Item model in LearningContextVersionItemVersion, so we can force that
+    # constraint?
+    published_version = models.ForeignKey('ItemVersion', on_delete=models.RESTRICT, null=True, related_name="+")
+    first_published_at = models.DateTimeField(null=True)  # can be null if never published 
+    last_published_at = models.DateTimeField(null=True)   # can be null if never published
+
+    created = manual_date_time_field()
+    modified = manual_date_time_field()
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["learning_context_id", "identifier"],
+                name="learning_publishing_lcb_one_identifier_per_lc",
+            )
+        ]
+
+    def __str__(self):
+        return f"{self.identifier}"
+
+
+class ItemVersion(models.Model):
+    """
+    A particular version of an Item.
+
+    A new ItemVersion is created anytime there is either a change to the content
+    or a change to the policy around a piece of content (e.g. schedule change).
+    """
+    uuid = immutable_uuid_field()
+    item_info = models.ForeignKey('ItemInfo', on_delete=models.RESTRICT)
+    item = models.ForeignKey(Item, on_delete=models.CASCADE)
+    title = models.CharField(max_length=1000, blank=True, null=True)
+
+    def __str__(self):
+        return f"{self.uuid}: {self.title}"
+
+
+class LearningContextVersionItemVersion(models.Model):
+    learning_context_version = models.ForeignKey(LearningContextVersion, on_delete=models.CASCADE)
+    item_version = models.ForeignKey(ItemVersion, on_delete=models.RESTRICT)
+
+
+class ItemInfo(models.Model):
+    """
+    Basic metadata about an Item.
+
+    This is mostly here to give supplmentary models to hang off of this and give
+    more interesting models based on the data in ItemRaw.
+    """
+    uuid = immutable_uuid_field()
+    learning_context = models.ForeignKey(LearningContext, on_delete=models.CASCADE)
+    type = models.CharField(max_length=100)
+    item_raw = models.ForeignKey('ItemRaw', on_delete=models.RESTRICT)
+
+    created = manual_date_time_field()
+
+
+class ItemRaw(models.Model):
+    """
+    This is the most basic piece of raw Item content data, with no metadata.
+
+    ItemRaw stores data in a Binary BLOB `data` field. This data is not
+    normalized in any way, meaning that pieces of content that are semantically
+    equivalent (e.g. differently spaced/sorted JSON) will result in new entries.
+    This model is intentionally ignorant of what these things mean, because it
+    expects supplemental data models to build on top of it.
+
+    The other fields on ItemRaw are for data that is intrinsic to the file data
+    itself (e.g. the size). Any smart parsing of the contents into more
+    structured metadata should happen in other models that hang off of ItemInfo.
```

Comment:
@ormsbee I am struggling a bit with the distinction between ItemInfo and ItemRaw.

What would we lose by collapsing them into a single model? Is there something bad about structured metadata models hanging off of the same model that's storing the raw bytes?

Reply:
I think I originally created the idea of this model (under a different name), when I was thinking it would have more stuff in it (like a local identifier associated with the content package), but those concepts were simplified out of the design. We would have to be careful to not pull back the raw payload all the time, but that's manageable.

The only remaining use case I can think of is when you want to add a little bit of metadata about the content that is not the content itself–new information entered by a human, not something that's being derived by the code from an `ItemRaw`. At that point, we have to figure out where to hang that data in a way that also clearly signals that this content has changed.

As an example, let's say we have a text description of an image.

*If we attach that data to a model that hangs off of `ItemRaw`*, then there's no indication that the content has been updated. None of the versioning mechanisms above this were invoked or needed to be invoked. So we've done an unexpected mutation, and it will invalidate assumptions around versioning and caching. We can't even create a new `ItemRaw` to bind it to (wasteful as that would be in terms of space), because `ItemRaw` uses a hash of its contents for identity.

*We could attach that data to a model that hangs off of `ItemVersion`.*  That's a little weird because that's not policy data that changes from course to course, but we would have default policy information that's attached at that layer and copied, so maybe it would work out...?

You're right in that other comment, it's very much like the definition/usage split in the old system. But maybe we should take that as a cautionary tale–people put many things in the settings scope that should have been put in definitions. If the distinction of when to use `ItemInfo` is unclear to you as an expert in this kind of system, then it will almost certainly confuse others who try to use this later. Maybe it's easier to say: "If you can completely derive it from the raw data, hang a model off of `ItemRaw`, and if it's any new data entered in by a user, hang it off of `ItemVersion`."?

Reply:
I just realized I need to check an assumption about how I've been reading this: "hanging model X off of model Y" means that X would have a foreign key to Y, right?